File: android/expoview/src/main/java/host/exp/exponent/kernel/ExponentErrorMessage.java
// Copyright 2015-present 650 Industries. All rights reserved.

package host.exp.exponent.kernel;

public class ExponentErrorMessage {

  private static final int MAX_LENGTH = 300;

  private String mUserErrorMessage;
  private final String mDeveloperErrorMessage;

  public ExponentErrorMessage(String userErrorMessage, String developerErrorMessage) {
    this.mUserErrorMessage = userErrorMessage;
    this.mDeveloperErrorMessage = developerErrorMessage;
  }

  public String userErrorMessage() {
    if (mUserErrorMessage != null) {
      return limit(mUserErrorMessage);
    } else {
      return "";
    }
  }

  public String developerErrorMessage() {
    if (mDeveloperErrorMessage != null) {
      return limit(mDeveloperErrorMessage);
    } else {
      return "";
    }
  }

  public static ExponentErrorMessage userErrorMessage(final String errorMessage) {
    return new ExponentErrorMessage(errorMessage, errorMessage);
  }

  public static ExponentErrorMessage developerErrorMessage(final String errorMessage) {
    return new ExponentErrorMessage(null, errorMessage);
  }

  public ExponentErrorMessage addUserErrorMessage(final String errorMessage) {
    mUserErrorMessage = errorMessage;
    return this;
  }

  private String limit(final String s) {
    if (s.length() < MAX_LENGTH) {
      return s;
    } else {
      return s.substring(0, MAX_LENGTH);
    }
  }
}

File: android/expoview/src/main/java/host/exp/exponent/kernel/ExponentErrorMessage.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package host.exp.exponent.kernel

private const val MAX_LENGTH = 600

class ExponentErrorMessage(
  private var userErrorMessage: String?,
  private val developerErrorMessage: String?
) {
  fun userErrorMessage(): String {
    return userErrorMessage?.let { limit(it) } ?: ""
  }

  fun developerErrorMessage(): String {
    return developerErrorMessage?.let { limit(it) } ?: ""
  }

  fun addUserErrorMessage(errorMessage: String?): ExponentErrorMessage {
    userErrorMessage = errorMessage
    return this
  }

  private fun limit(s: String): String {
    return if (s.length < MAX_LENGTH) {
      s
    } else {
      s.substring(0, MAX_LENGTH)
    }
  }

  companion object {
    @JvmStatic fun userErrorMessage(errorMessage: String?): ExponentErrorMessage {
      return ExponentErrorMessage(errorMessage, errorMessage)
    }

    @JvmStatic fun developerErrorMessage(errorMessage: String?): ExponentErrorMessage {
      return ExponentErrorMessage(null, errorMessage)
    }
  }
}
